backend/corpora/common/dataset_validator.py

```diff
@@ -40,21 +43,85 @@ def validate_dataset_file(self):
 
     def validate_h5ad_dataset(self, file_object):
         """
-        Reads the file contents into an AnnData object. Each attribute of the AnnData object will then be checked to
-        ensure it contains the appropriate metadata.
+        Reads the H5AD file contents into an AnnData object. Each attribute of the AnnData object will then be
+        checked to ensure it contains the appropriate metadata.
         """
 
         start_time = time.time()
-        logging.info("Reading anndata object...")
+        logging.info("Reading H5AD file into anndata object...")
         anndata_object = anndata.read_h5ad(file_object)
         logging.info(f"Finished reading anndata object in {time.time() - start_time:.3f} seconds.")
 
+        self.validate_anndata_object(anndata_object)
+
+    def validate_loom_dataset(self, file_object):
+        """
+        Reads the Loom file contents into an AnnData object. Each attribute of the AnnData object will then be
+        checked to ensure it contains the appropriate metadata.
+        """
+
         start_time = time.time()
-        logging.info("Beginning validation of file...")
+        logging.info("Reading Loom file into anndata object...")
+        anndata_object = anndata.read_loom(file_object)
+        logging.info(f"Finished reading anndata object in {time.time() - start_time:.3f} seconds.")
+
+        self.validate_anndata_object(anndata_object)
+
+    def validate_anndata_object(self, anndata_object: anndata.AnnData):
+        start_time = time.time()
+        logging.info("Beginning validation of anndata object...")
+        self.verify_layers(anndata_object)
         self.verify_obs(anndata_object)
         self.verify_vars(anndata_object)
         self.verify_uns(anndata_object)
-        logging.info(f"Finished completing validation on the file in {time.time() - start_time:.3f} seconds.")
+        logging.info(f"Finished completing validation in {time.time() - start_time:.3f} seconds.")
+
+    def verify_layers(self, data_object: anndata.AnnData):
+        """
+        Verifies that the dataset contains at least the raw data and if other layers are provided, that they each
+        contain an appropriate description.
+        """
+
+        # Check to make sure X data exists
+        data = None
+        if isinstance(data_object.X, DataFrame):
+            data = data_object.X.data
+        elif isinstance(data_object.X, ndarray):
+            data = data_object.X
+        elif isinstance(data_object.X, csr_matrix):
+            data = data_object.X.toarray()
```

Comment:
This might blow up the memory, no? There are some big datasets around that are like 1% nonzero.

Reply:
Switched to use count_non_zero compared to the number of total elements. Hopefully this is less memory intensive.